libcrux-ml-kem/src/vector/portable/serialize.rs

```diff
@@ -4,321 +4,403 @@ use crate::vector::traits::FIELD_ELEMENTS_IN_VECTOR;
-    result.elements[2] = ((bytes[3] as i16 & 0x3F) << 4 | (bytes[2] as i16 >> 4)) as i16;
-    result.elements[3] = (((bytes[4] as i16) << 2) | (bytes[3] as i16 >> 6)) as i16;
-    result.elements[4] = ((bytes[6] as i16 & 0x03) << 8 | (bytes[5] as i16 & 0xFF)) as i16;
-    result.elements[5] = ((bytes[7] as i16 & 0x0F) << 6 | (bytes[6] as i16 >> 2)) as i16;
-    result.elements[6] = ((bytes[8] as i16 & 0x3F) << 4 | (bytes[7] as i16 >> 4)) as i16;
-    result.elements[7] = (((bytes[9] as i16) << 2) | (bytes[8] as i16 >> 6)) as i16;
 
-    result.elements[8] =
-        ((bytes[10 + 1] as i16 & 0x03) << 8 | (bytes[10 + 0] as i16 & 0xFF)) as i16;
-    result.elements[9] = ((bytes[10 + 2] as i16 & 0x0F) << 6 | (bytes[10 + 1] as i16 >> 2)) as i16;
-    result.elements[10] = ((bytes[10 + 3] as i16 & 0x3F) << 4 | (bytes[10 + 2] as i16 >> 4)) as i16;
-    result.elements[11] = (((bytes[10 + 4] as i16) << 2) | (bytes[10 + 3] as i16 >> 6)) as i16;
-    result.elements[12] =
-        ((bytes[10 + 6] as i16 & 0x03) << 8 | (bytes[10 + 5] as i16 & 0xFF)) as i16;
-    result.elements[13] = ((bytes[10 + 7] as i16 & 0x0F) << 6 | (bytes[10 + 6] as i16 >> 2)) as i16;
-    result.elements[14] = ((bytes[10 + 8] as i16 & 0x3F) << 4 | (bytes[10 + 7] as i16 >> 4)) as i16;
-    result.elements[15] = (((bytes[10 + 9] as i16) << 2) | (bytes[10 + 8] as i16 >> 6)) as i16;
+#[inline(always)]
+pub(crate) fn deserialize_10(bytes: &[u8]) -> PortableVector {
+    let v0_7 = deserialize_10_int(&bytes[0..10]);
+    let v8_15 = deserialize_10_int(&bytes[10..20]);
+    let mut v = zero();
+    v.elements[0] = v0_7.0;
+    v.elements[1] = v0_7.1;
+    v.elements[2] = v0_7.2;
+    v.elements[3] = v0_7.3;
+    v.elements[4] = v0_7.4;
+    v.elements[5] = v0_7.5;
+    v.elements[6] = v0_7.6;
+    v.elements[7] = v0_7.7;
+    v.elements[8] = v8_15.0;
+    v.elements[9] = v8_15.1;
+    v.elements[10] = v8_15.2;
+    v.elements[11] = v8_15.3;
+    v.elements[12] = v8_15.4;
+    v.elements[13] = v8_15.5;
+    v.elements[14] = v8_15.6;
+    v.elements[15] = v8_15.7;
+    v
+}
 
-    result
+#[inline(always)]
+pub(crate) fn serialize_11_int(v: &[i16]) -> (u8, u8, u8, u8, u8, u8, u8, u8, u8, u8, u8) {
+    let r0 =   v[0] as u8;
+    let r1 = ((v[1] & 0x1F) as u8) << 3 | ((v[0] >> 8) as u8);
+    let r2 = ((v[2] & 0x3) as u8) << 6 | ((v[1] >> 5) as u8);
+    let r3 = ((v[2] >> 2) & 0xFF) as u8;
+    let r4 = ((v[3] & 0x7F) as u8) << 1 | (v[2] >> 10) as u8;
+    let r5 = ((v[4] & 0xF) as u8) << 4 | (v[3] >> 7) as u8;
+    let r6 = ((v[5] & 0x1) as u8) << 7 | (v[4] >> 4) as u8;
+    let r7 = ((v[5] >> 1) & 0xFF) as u8;
+    let r8 = ((v[6] & 0x3F) as u8) << 2 | (v[5] >> 9) as u8;
+    let r9 = ((v[7] & 0x7) as u8) << 5 | (v[6] >> 6) as u8;
+    let r10 = (v[7] >> 3) as u8;
+    (r0,r1,r2,r3,r4,r5,r6,r7,r8,r9,r10)
 }
 
 #[inline(always)]
 pub(crate) fn serialize_11(v: PortableVector) -> [u8; 22] {
+    let r0_10 = serialize_11_int(&v.elements[0..8]);
```

Comment:
Is there a particular reason this style of splitting the serialize functions was adopted?

Reply:
I added a comment on top of this file. 
```
//! A general style adopted here is to first define an internal function
//! called serialize_N_int or deserialize_N_int that (de)serializes
//! the minimal number of inputs K such that N*K is a multiple of 8.
//! These functions are then called multiple times in the main function,
//! called serialize_N or deserialize_N.
//! This factoring reduces redundancy, and also makes he code easier for
//! F* to handle. As a general rule, any function that modifes an array
//! more than 8 times with complex expressions starts to strain F*, so
//! we separate out the code that does the computation (in _int functions)
//! and code that updates arrays (in the outer functions).
```

Reply:
More specifically, F* works very well if we do the main computation in functions that return tuples and then only do simple array updates in the outer functions, hence this style choice. Many of these functions would also work if we used arrays everywhere, but they would be slow to even lax-check, which would then make full verification even slower.

Reply:
I understand the reason. But it's pretty ugly. We should try to find a better style later.

Reply:
There are two things here.

The first is breaking down the functions into bits that repeat. This is a "good thing" for both code clarity and verification. Indeed, our original code in `src/kyber` was written this way. 

E.g. see the loop body in https://github.com/cryspen/libcrux/blob/02472265ddcd5af5cb1a9778855e081cb7f083b5/libcrux-ml-kem/src/kem/kyber/serialize.rs#L195 

It also breaks the input into chunks of 8, and here we are doing the same for a vector with 16 chunks.
The operations on the two chunks of 8 is the same.
I do not think this is an uglification, and it should not be changed.

The second bit is the use of tuples in the internal functions, and this is indeed an artifact of making it work in F*.
If we can, we should switch this to maybe working on arrays or mutable slices or something more idiomatic in Rust.

Reply:
I also added a comment to the code. There is another thing I tried which is to return an array expression instead of copying the tuple to the array in a long sequence of array updates. This was slowing down F* but still terminating, but I think when we start doing proofs, this slowdown will hurt more. So I leave it for a future pass where we prettify the code without killing F*.